### PR TITLE
EVM

### PR DESCRIPTION
**Pull request by mpapierski**
## Summary

This PR adds a proof-of-concept EVM execution path to `casper-node`. It proves that Casper can accept, validate, execute, account for, and persist Ethereum-shaped transactions while keeping the execution and storage model rooted in Casper: global state, transaction flow, account/purse accounting, fee/refund policy.

The companion `casper-sidecar` changes in casper-network/casper-sidecar#468 provide the Ethereum JSON-RPC surface for this PoC. Sidecar currently implements `eth_chainId`, `eth_blockNumber`, `eth_getBlockByNumber`, `eth_getTransactionCount`, `eth_sendRawTransaction`, `eth_getTransactionReceipt`, `eth_getLogs`, `eth_newFilter`, `eth_getFilterChanges`, `eth_getFilterLogs`, `eth_uninstallFilter`, `eth_call`, and websocket log subscriptions. It uses node binary-port APIs to submit signed EVM transactions, run speculative EVM calls, read blocks and transaction execution info, and project Casper EVM receipts/logs into Ethereum-compatible JSON-RPC responses.

## What Changed

On the node side, this branch introduces:

- `Transaction::Evm` and `TransactionHash::Evm`.
- Casper-owned EVM types for addresses, hashes, transactions, accounts, config, receipts, logs, topics, and Ethereum-style `U256` JSON quantities.
- A new `casper-executor-evm` crate backed by `revm`.
- EVM global-state keys for account identity, nonce, code hash, bytecode, and storage.
- `ExecutionResult::Evm`, carrying EVM receipt data plus Casper accounting fields.
- Transaction acceptor support for EVM enablement, signature validation, chain ID, gas limit, base fee, priority fee, balance, and nonce checks.
- Contract-runtime routing for finalized EVM transactions.
- Casper fee/refund handling for EVM transactions.
- Binary-port speculative execution support for read-only `eth_call`.
- Native Casper transfers to 20-byte EVM addresses when EVM is enabled.
- EIP-7702 set-code transaction support.

## What Works

The current PoC supports signed Ethereum RLP ingestion through sidecar, conversion into `Transaction::Evm`, node-side validation, execution through `revm`, persistence of EVM state, and Ethereum-compatible receipt projection.

Tested flows include contract deployment, contract calls, native EVM value transfer, event logs in receipts, simple ERC20/ERC721-style behavior, speculative `eth_call`, native Casper funding of EVM addresses, and EIP-7702 delegated-code execution using Foundry.

The PoC supports the Ethereum transaction envelopes needed for the current tooling flow:

- Legacy transactions, with [EIP-155](https://eips.ethereum.org/EIPS/eip-155) chain ID replay protection.
- [EIP-2930](https://eips.ethereum.org/EIPS/eip-2930) access-list transactions, only when the access list is empty.
- [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) dynamic-fee transactions, only when `max_priority_fee_per_gas == 0`.
- [EIP-7702](https://eips.ethereum.org/EIPS/eip-7702) set-code transactions, with non-empty authorization lists, empty access lists, and zero priority fee.

## Prague/Pectra Compatibility

The executor is configured with a Prague-capable EVM spec, and this branch implements the main PoC execution path needed for [EIP-7702](https://eips.ethereum.org/EIPS/eip-7702). In that sense, the exercised EVM path is Prague-compatible.

It is not yet a complete Prague/Electra network implementation. The canonical Prague/Electra meta EIP is [EIP-7600](https://eips.ethereum.org/EIPS/eip-7600). Remaining Prague/Pectra work includes:

- [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537): BLS12-381 precompiles.
- [EIP-2935](https://eips.ethereum.org/EIPS/eip-2935): historical block hash system contract.
- [EIP-6110](https://eips.ethereum.org/EIPS/eip-6110): validator deposit requests.
- [EIP-7002](https://eips.ethereum.org/EIPS/eip-7002): withdrawal request predeploy.
- [EIP-7251](https://eips.ethereum.org/EIPS/eip-7251): consolidation request predeploy.
- [EIP-7623](https://eips.ethereum.org/EIPS/eip-7623): calldata floor-cost accounting.
- [EIP-7685](https://eips.ethereum.org/EIPS/eip-7685): execution-layer requests and `requests_hash`.
- [EIP-7691](https://eips.ethereum.org/EIPS/eip-7691): Prague/Electra blob throughput parameters.
- [EIP-7549](https://eips.ethereum.org/EIPS/eip-7549): consensus-layer change; no execution-layer work expected here.

Blob transactions are still rejected. Full blob support requires [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) sidecars, KZG data, blob gas accounting, and the Prague blob parameter updates from EIP-7691. Client configuration should also eventually account for [EIP-7840](https://eips.ethereum.org/EIPS/eip-7840).

## Current Limitations

EVM is disabled by default in chainspecs and must be explicitly enabled.

The sidecar RPC surface is enough for the PoC Foundry flow, transaction submission, receipts, calls, logs, filters, and log subscriptions. Some Ethereum RPC methods are still future work, including `eth_estimateGas`, `eth_getBalance`, `eth_getCode`, and `eth_getTransactionByHash`.

`eth_getBlockByNumber` currently returns transaction hashes only; full transaction objects are explicitly rejected for now. Block and receipt projection still needs to mature around complete Ethereum roots, bloom handling, gas-used accounting, and historical query performance.

Access lists are decoded but non-empty lists are rejected. Priority fees are also rejected because Casper does not currently prioritize transactions using Ethereum fee-market tips.

## Future Work

Next steps are to move this from PoC towards a spec-complete feature:

- Add Prague/Pectra predeploy and system-contract handling.
- Implement EIP-7685 request collection and block/header commitments.
- Add blob transaction support, blob gas accounting, and blob schedule config.
- Complete EIP-7623 calldata floor-cost accounting.
- Define the long-term Casper policy for EVM base fee movement, priority fees, refunds, and fee distribution.
- Add indexed EVM logs for efficient historical queries.
- Expand sidecar RPC coverage and remove local patching once node-side types are released.
- Add conformance tests for Prague precompiles, EIP-7702 edge cases, blob rejection/support, and gas accounting boundaries.

Not every Ethereum feature maps cleanly onto Casper’s existing execution, storage, accounting, and consensus model. For those cases, the plan is to make a best-effort compatibility layer, keep the Casper-side semantics explicit, and document where behavior intentionally differs from Ethereum mainnet.

## Validation

This branch adds unit and integration coverage for EVM transaction decoding, approval handling, executor behavior, account/state storage, receipts, native transfers to EVM addresses, nonce validation, binary-port speculative EVM calls, Casper fee/refund handling, and EIP-7702 delegation behavior.

The manual Foundry validation flow is documented in `EVM.md`.

### CODE REVIEW

File: binary_port/src/speculative_execution_result.rs [hunk 1]
```diff
@@ -170,6 +171,128 @@ impl FromBytes for SpeculativeExecutionResult {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub struct EvmSpeculativeExecutionResult {
```
Comment (zajko) [inline binary_port/src/speculative_execution_result.rs:175]: Do we really need another struct for that? I think we could add "output: Bytes" to the existing `struct SpeculativeExecutionResult`, right?

File: execution_engine/src/engine_state/mod.rs [hunk 2]
```diff
@@ -76,7 +76,9 @@ impl ExecutionEngineV1 {
         // A good deal of effort has been put into removing all such behaviors; please do not
         // come along and start adding it back.
 
-        let account_hash = initiator_addr.account_hash();
+        let account_hash = initiator_addr
+            .account_hash()
+            .expect("Wasm v1 initiator must be a Casper account");
```
Comment (zajko) [inline execution_engine/src/engine_state/mod.rs]: But if this is an evm account backed by a casper account this still should be doable, right?

File: types/src/transaction/transfer_target.rs [hunk 3]
```diff
@@ -12,6 +12,8 @@ pub enum TransferTarget {
     PublicKey(PublicKey),
     /// An account hash.
     AccountHash(AccountHash),
+    /// An EVM address.
+    EvmAddress(evm::Address),
```
Comment (zajko) [inline types/src/transaction/transfer_target.rs:16]: Why do we need a separate transfer target for this? I thought that this will ultimately be resolved to either an account hash or a purse uref?

Comment (zajko) [inline types/src/transaction/transfer_target.rs:16]: the "random" function doesn't include the new variant

File: executor/evm/src/executor.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    };
+
+    let reimbursed_gas = gas_limit
+        .saturating_sub(gas.total_gas_spent())
+        .saturating_add(gas.inner_refunded());
+    let caller_reimbursement = U256::from(effective_gas_price) * U256::from(reimbursed_gas);
+    // Revm still computes Ethereum fee transfers internally before we remove
+    // them for Casper-owned fee accounting. For node-accepted EIP-1559
+    // transactions the priority fee is zero by policy, but this stays generic
+    // for executor callers and pre-London specs.
```
Comment (zajko) [inline executor/evm/src/executor.rs]: Should we be concerned with logic that supports transactions that were valid a long time ago? We won't have pre-Prague evm txes in our node, right?

File: types/src/evm/config.rs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+use crate::bytesrepr::{self, FromBytes, ToBytes, U8_SERIALIZED_LENGTH};
+
+/// Supported EVM hardfork specifications for chainspec configuration.
+#[derive(
+    Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Default, Serialize, Deserialize,
+)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+#[serde(rename_all = "snake_case")]
+pub enum EvmSpec {
```
Comment (zajko) [inline types/src/evm/config.rs:18]: Should we really have visibility of all these old specs? Couldn't we have only "Prague" here and add new ones once we support it?

File: node/src/components/network/tasks.rs [hunk 6]
```diff
@@ -685,111 +685,110 @@ where
     let demands_in_flight = Arc::new(Semaphore::new(context.max_in_flight_demands));
     let event_queue = context.event_queue.expect("component not initialized");
 
-    let read_messages = async move {
```
Comment (zajko) [inline node/src/components/network/tasks.rs:688]: are changes to this file necessary? I'm not sure if it's just an indentation refactor or did some logic actually change

File: node/src/components/transaction_acceptor/error.rs [hunk 7]
```diff
@@ -115,6 +115,9 @@ impl From<Error> for BinaryPortErrorCode {
                     DeployParameterFailure::MissingTransferTarget => {
                         BinaryPortErrorCode::DeployMissingTransferTarget
                     }
+                    DeployParameterFailure::EvmAddressTransferDisabled => {
```
Comment (zajko) [inline node/src/components/transaction_acceptor/error.rs:118]: Shouldn't we also be rejecting Transaction::Evm, InitiatorAddr::Evm and everything evm-related if evm_config.enabled is false?

File: node/src/components/contract_runtime/operations.rs [hunk 8]
```diff
@@ -209,18 +543,24 @@ pub fn execute_finalized_block(
...
-        let transaction_args = transaction.session_args().clone();
-        let entry_point = transaction.entry_point();
-        let authorization_keys = transaction.signers();
+        let initiator_addr = stored_transaction.initiator_addr();
+        let transaction_hash = stored_transaction.hash();
+        let authorization_keys = stored_transaction.authorization_keys();
+
+        if !is_evm {
+            if let Some(address) = initiator_addr.evm_address() {
+                return Err(BlockExecutionError::EvmInitiatorForNonEvmTransaction { address });
```
Comment (zajko) [inline node/src/components/contract_runtime/operations.rs]: It's currently legal to create a Transaction::V1 with InitiatiorAddr::Evm and fail the block. I actually think that it will kill the whole node, the error msg is 
```
2026-05-25T22:06:48.065104Z ERROR [casper_node::components::contract_runtime::utils utils.rs:358] failed to execute block; error=EVM initiator address Address([25, 231, 227, 118, 231, 194, 19, 183, 231, 231, 228, 108, 199, 10, 93, 208, 134, 218, 255, 42]) is only valid for EVM transactions
```
and the node exits

File: types/src/transaction/initiator_addr.rs [hunk 9]
```diff
@@ -28,37 +29,59 @@ const PUBLIC_KEY_FIELD_INDEX: u16 = 1;
...
+    schemars(description = "The address of the initiator of a transaction.")
 )]
 #[serde(deny_unknown_fields)]
 pub enum InitiatorAddr {
     /// The public key of the initiator.
     PublicKey(PublicKey),
     /// The account hash derived from the public key of the initiator.
     AccountHash(AccountHash),
+    /// The EVM-native address recovered from the signed Ethereum transaction.
+    EvmAddress(evm::Address),
```
Comment (zajko) [inline types/src/transaction/initiator_addr.rs]: Introducing another InitiatorAddr variant which is applicable only with some of the transactions (but structurally fits to V1) is problematic. I think we should not be doing that.

File: types/src/evm/transaction.rs [hunk 10]
```diff
@@ -0,0 +1,1510 @@
...
+            ));
+        }
+        if envelope
+            .access_list()
+            .is_some_and(|access_list| !access_list.is_empty())
+        {
+            return Err(TransactionError::UnsupportedAccessList);
+        }
+
+        let kind = if envelope.is_legacy() {
```
Comment (zajko) [inline types/src/evm/transaction.rs:908]: Can we have no expectation that the new transactions introduced to our evm env will be somehow contemporary? Other evm compliant chains also support all the legacy kinds and formats?

File: node/src/components/transaction_acceptor.rs [hunk 11]
```diff
@@ -211,11 +293,30 @@ impl TransactionAcceptor {
...
+
         if event_metadata.source.is_client() {
-            let account_hash = match event_metadata.transaction.initiator_addr() {
-                InitiatorAddr::PublicKey(public_key) => public_key.to_account_hash(),
-                InitiatorAddr::AccountHash(account_hash) => account_hash,
-            };
+            let initiator_addr = event_metadata.transaction.initiator_addr();
+            let account_hash = initiator_addr
+                .account_hash()
+                .expect("non-EVM transaction initiator must be a Casper account");
```
Comment (zajko) [inline node/src/components/transaction_acceptor.rs]: expect panics, right?
So if someone sends a V1 transaction with InitiatorAddr::EVM that will crash the node?

File: node/src/components/transaction_acceptor.rs
Comment (zajko) [inline node/src/components/transaction_acceptor.rs:1]: I have a feeling that we need more checks here.

File: node/src/types/transaction/meta_transaction/meta_evm.rs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                // dynamic-fee transactions without implying transaction
+                // priority based on gas parameters.
+                let max_fee_per_gas = transaction.max_fee_per_gas();
+                if max_fee_per_gas < base_fee {
+                    return Err(evm::TransactionError::MaxFeePerGasBelowBaseFee {
+                        max_fee_per_gas,
+                        base_fee,
+                    });
+                }
+                let max_priority_fee_per_gas = transaction.max_priority_fee_per_gas().unwrap_or(0);
```
Comment (zajko) [inline node/src/types/transaction/meta_transaction/meta_evm.rs:144]: What handles the "max_priority_fee_per_gas"? I don't think it's evm execution engine - it's probably the node, right? If yes, why would we fail because of this? Isn't it true that any number here would be just ignored by casper node? We wouldn't give anyone any priority but we would also not apply any priority fee so no harm done, right?

File: smart_contracts/contract/src/no_std_handlers.rs [hunk 13]
```diff
@@ -2,7 +2,6 @@
 
 /// A panic handler for use in a `no_std` environment which simply aborts the process.
 #[panic_handler]
-#[no_mangle]
```
Comment (zajko) [inline smart_contracts/contract/src/no_std_handlers.rs]: Why did this change? A bit surprising

File: storage/src/global_state/state/mod.rs [hunk 14]
```diff
@@ -1489,7 +1492,7 @@ pub trait StateProvider: Send + Sync + Sized {
                 // pay amount from source to target
                 match runtime
                     .transfer(
-                        Some(initiator_addr.account_hash()),
+                        initiator_addr.account_hash(),
```
Comment (zajko) [inline storage/src/global_state/state/mod.rs]: I don't think `None` meant the same thing as it does now

File: types/src/block/era_end.rs [hunk 15]
```diff
@@ -61,7 +61,7 @@ impl EraEnd {
     }
 
     /// Returns the rewards.
-    pub fn rewards(&self) -> Rewards {
+    pub fn rewards(&self) -> Rewards<'_> {
```
Comment (zajko) [inline types/src/block/era_end.rs:64]: This is a rust-toolchain bump requirement?

File: types/src/evm/eth_u256.rs
Comment (zajko) [inline types/src/evm/eth_u256.rs:1]: does this have to be part of the public api? I see it only being used as a `pub use` in evm.rs

File: types/src/evm/transaction.rs [hunk 16]
```diff
@@ -0,0 +1,1510 @@
...
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for TransactionError {}
+
+/// An unsigned Ethereum transaction payload plus one Ethereum-style Casper approval.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+pub struct Transaction {
```
Comment (zajko) [inline types/src/evm/transaction.rs]: Not sure if we want to call it a "transaction" since it has such a load-bearing meaning in the node

File: types/src/byte_code.rs [hunk 17]
```diff
@@ -420,6 +422,20 @@ pub enum ByteCodeKind {
     V1CasperWasm = 1,
     /// Byte code to be executed with the version 2 Casper execution engine.
     V2CasperWasm = 2,
+    /// Prague-compatible EVM bytecode.
+    ///
+    /// This variant records bytecode that is valid for the Prague EVM rules.
+    /// When support for a future bytecode-affecting EVM spec such as Osaka is
+    /// officially added, introduce a new `Evm<Spec>` variant instead of
+    /// changing the meaning of this one.
+    EvmPrague = 3,
```
Comment (zajko) [inline types/src/byte_code.rs:431]: So inherently if someone gives us a specid that is pre-prague we should be rejecting it, right?

File: binary_port/src/speculative_execution_result.rs (hunk 1 above)
Comment (mpapierski) [reply binary_port/src/speculative_execution_result.rs:175]: Fair point for output bytes, although for VM1, it's not raw bytes, it's `CLValue`. Fitting raw output bytes from EVM into CLValue feels weird. There's also `evm_receipt` in there, which is very EVM-specific.

File: types/src/transaction/transfer_target.rs (hunk 3 above)
Comment (mpapierski) [reply types/src/transaction/transfer_target.rs:16]: The idea here is that a user with a Casper identity (aka `Key::Account`) can seed EVM addresses via native transfers. Think of exchanges, wallets, etc, being able to send funds to 20-byte addresses using their funds from existing hot wallets.

File: types/src/block/era_end.rs (hunk 15 above)
Comment (mpapierski) [reply types/src/block/era_end.rs:64]: had to bump it to ~1.91.1~ 1.91.0 to make revm happy, but as you can see in the Makefile, I added a lot of exemptions to keep the diff as small as possible. Otherwise, it will be gigantic. I intend to address these exemptions in follow up PR.

File: storage/src/global_state/state/mod.rs (hunk 14 above)
Comment (mpapierski) [reply storage/src/global_state/state/mod.rs]: As implemented in this branch, `fn account_hash()` returned Optional, so technically it'd be the same (Some would be guaranteed by earlier code branches for AccountHash/PublicKey). I'm working on removing `Evm` from `InitiatorAddr` enum to simplify the code and minimize diff against dev branch.

File: node/src/types/transaction/meta_transaction/meta_evm.rs (hunk 12 above)
Comment (mpapierski) [reply node/src/types/transaction/meta_transaction/meta_evm.rs:144]: For now, we will not implement priority-fee ordering or payment, since transactions are neither ordered by fee nor prioritized in any way. The current approach is to accept these fields in the serialized ETH transaction format, but have the transaction acceptor reject any non-zero values. This avoids silently ignoring them, which could otherwise be confusing for ETH users.

File: types/src/transaction/initiator_addr.rs (hunk 9 above)
Comment (mpapierski) [reply types/src/transaction/initiator_addr.rs]: Agreed it causes more problems than it solves in 0b52027cd56665eb871ba2a024822f63b412ebb2

File: node/src/components/transaction_acceptor/error.rs (hunk 7 above)
Comment (mpapierski) [reply node/src/components/transaction_acceptor/error.rs:118]: `InitiatorAddr::Evm` removed in 0b52027cd5. We do both reject transfers in the transaction acceptor if `evm.enabled == false` and additionally, a native transfer to a 20-byte address is rejected when `evm.enabled == false`. That flag also applies to speculative execution.

File: execution_engine/src/engine_state/mod.rs (hunk 2 above)
Comment (mpapierski) [reply execution_engine/src/engine_state/mod.rs]: Addressed by 0b52027cd5

File: node/src/components/contract_runtime/operations.rs (hunk 8 above)
Comment (mpapierski) [reply node/src/components/contract_runtime/operations.rs]: Addressed by 0b52027cd5

File: node/src/components/transaction_acceptor.rs (hunk 11 above)
Comment (mpapierski) [reply node/src/components/transaction_acceptor.rs]: Same - addressed by 0b52027cd5

File: smart_contracts/contract/src/no_std_handlers.rs (hunk 13 above)
Comment (mpapierski) [reply smart_contracts/contract/src/no_std_handlers.rs]: Compiler behavior changed on 1.91.0 - `E0554: '#[no_mangle]' cannot be used on internal language items`

File: types/src/evm/transaction.rs (hunk 16 above)
Comment (mpapierski) [reply types/src/evm/transaction.rs]: Addressed in 9d23df34774c5b3d821ce19a8e5c07de2e0a35fb

File: types/src/byte_code.rs (hunk 17 above)
Comment (mpapierski) [reply types/src/byte_code.rs:431]: Addressed in 8818352f81 - removed older pre-prague specs so it's impossible to setup revm with pre-prague spec with prague features we support right now.

File: executor/evm/src/executor.rs (hunk 4 above)
Comment (mpapierski) [reply executor/evm/src/executor.rs]: The pre-London branches (or pre-Prague) are gone in 8818352f81. We do however have to check the configs and instantiate revm with the appropriate SpecId. On executor side we're pretty much generic over what kind of transaction comes, so the validation of all supported "transaction kinds" happens in the node itself.

File: types/src/evm/config.rs (hunk 5 above)
Comment (mpapierski) [reply types/src/evm/config.rs:18]: Should be gone in 8818352f818121e4bdd4cd361e55c4889ca425ef  -we'd only list specs we support onwards, starting with Prague

File: node/src/components/network/tasks.rs (hunk 6 above)
Comment (mpapierski) [reply node/src/components/network/tasks.rs:688]: Happened around toolchain update - reverted

File: types/src/transaction/transfer_target.rs (hunk 3 above)
Comment (mpapierski) [reply types/src/transaction/transfer_target.rs:16]: Addressed in 8818352f81

File: types/src/evm/transaction.rs (hunk 10 above)
Comment (mpapierski) [reply types/src/evm/transaction.rs:908]: Other way around - we do accept all four common envelope formats: Legacy (with EIP-155 chain ID), EIP-2930, EIP-1559, and EIP-7702. The `from_signed_rlp` match dispatches to all of them, and the PR description enumerates the same list.

The PoC restrictions are narrower than "no legacy support":

EIP-2930 / EIP-7702: access list must be empty (the executor doesn't model warm/cold access accounting yet; will be addressed in follow-up PRs)
EIP-1559 / EIP-7702: priority fee must be zero, since we don't order or prioritize transactions by tip - accepting non-zero would silently mislead users (same reasoning as the `NonZeroMaxPriorityFeePerGas` rejection you flagged in other comment)

The only envelope we explicitly reject is EIP-4844 (blob transactions) - that one needs KZG commitments, blob sidecars, blob base fee accounting, and block-header blob commitments, which is substantial work that we'd want to do as a follow-up alongside the rest of EIP-7691 (assuming we do want to support that).

File: node/src/components/transaction_acceptor.rs
Comment (mpapierski) [reply node/src/components/transaction_acceptor.rs:1]: For now, there are basic checks implemented (nonce, addresses, dynamic fees == 0, config compliance, etc.) I think we will figure out more valuable checks as we get more familiar with different ETH tooling and conventions.

File: types/src/evm/eth_u256.rs
Comment (mpapierski) [reply types/src/evm/eth_u256.rs:1]: I moved this type to sidecar, which has use there - the convention eth establishes is to prefix each human form representation with `0x{base16}` unlike our U256 type do (`base10`)

File: types/src/evm/transaction.rs [hunk 18]
```diff
@@ -0,0 +1,1513 @@
...
+            + self.max_priority_fee_per_gas.serialized_length()
+            + self.value.serialized_length()
+            + Bytes::from(self.input.clone()).serialized_length()
+            + self.chain_id.serialized_length()
+            + self.authorization_list.serialized_length()
+            + self.approvals.serialized_length()
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), bytesrepr::Error> {
+        self.timestamp.write_bytes(writer)?;
```
Comment (zajko) [inline types/src/evm/transaction.rs]: I think we could benefit from using calltable semantics here.
Also, since this data is evm-transaction plus some metadata we could potentially separate them? Just thinking out loud, maybe it would be good to have a field that we store as a binary blob that is the actual thing that was signed over?

File: types/src/evm/transaction.rs [hunk 19]
```diff
@@ -0,0 +1,1513 @@
...
+    }
+
+    fn approval_signature(
+        &self,
+        signature_hash: &B256,
+    ) -> Result<(AlloySignature, Address), EvmTransactionError> {
+        let approval = self.single_approval()?;
+        let raw_signature = secp256k1_signature_bytes(approval)?;
+        let expected_signer = approval.signer();
+        for y_parity in [false, true] {
```
Comment (zajko) [inline types/src/evm/transaction.rs]: As we disussed on the call - this looks fishy.

File: types/src/evm/transaction.rs [hunk 20]
```diff
@@ -0,0 +1,1513 @@
...
+        }
+        Err(EvmTransactionError::InvalidApprovalSignature)
+    }
+
+    fn single_approval(&self) -> Result<&Approval, EvmTransactionError> {
+        let mut approvals = self.approvals.iter();
+        let approval = approvals
+            .next()
+            .ok_or(EvmTransactionError::MissingApproval)?;
+        if approvals.next().is_some() {
```
Comment (zajko) [inline types/src/evm/transaction.rs]: Do we verify this in the transaction acceptor?

**Review by EdHastingsCasperAssociation — APPROVED** [review]
per convo yesterday, approved to merge to feature branch to allow further collab.
